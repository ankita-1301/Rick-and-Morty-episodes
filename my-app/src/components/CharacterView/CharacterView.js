import React from "react";
import "../../styles.css";
import "antd/dist/antd.css";
import { Card, Avatar } from "antd";
import { UserOutlined } from "@ant-design/icons";
import { Link } from "react-router-dom";

export default class CharacterView extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      episodes: [],
      recommendations: [],
      species: "",
    };
  }

  componentDidMount() {
    this.fetchCharacter();
    this.fetchRecommendations();
  }

  fetchCharacter = () => {
    const {
      location: {
        state: { data },
      },
    } = this.props;
    data.episode.map((episodePage) => {
      fetch(`${episodePage}`)
        .then((response) => response.json())
        .then((eachEpisode) => {
          this.setState({
            episodes: [...this.state.episodes, ...[eachEpisode]],
          });
        })
        .catch((err) => console.log(err));
    });
  };

  fetchRecommendations = () => {
    const {
      location: {
        state: {
          data: { species },
        },
      },
    } = this.props;

    fetch(`https://rickandmortyapi.com/api/character/?species=${species}`)
      .then((response) => response.json())
      .then(({ results }) => {
        results.map((character) => {
          character.episode.map((eachEpisode) => {
            fetch(`${eachEpisode}`)
              .then((response) => response.json())
              .then((episodeData) => {
                this.setState({
                  recommendations: [
                    ...this.state.recommendations,
                    ...[episodeData],
                  ],
                });
              })
              .catch((err) => console.log(err));
          });
        });
      })
      .catch((err) => console.log(err));
  };

  render() {
    const { episodes, recommendations } = this.state;
    const {
      location: {
        state: { data },
      },
    } = this.props;

    //remove duplicates in recommended episodes
    let relevantRecommends = Array.from(
      new Set(recommendations.map((a) => a.id))
    )
      .map((id) => recommendations.find((a) => a.id === id))
      .filter((o) => !episodes.find((o2) => o.id === o2.id));

    const gridStyle = {
      width: "25%",
      textAlign: "center",
      height: "120px",
      padding: "10px",
    };

    return (
      <div data-test="characterView-div">
        <h1>{data.name}</h1>
        <Avatar
          shape="square"
          size={200}
          icon={<UserOutlined />}
          src={data.image}
          className="character-avatar"
        />
        <h5> Status : {data.status}</h5>
        <h5> Species : {data.species}</h5>
        <h5> Gender : {data.gender}</h5>
        <h5> Type : {data.type}</h5>
        <Card
          title="Episodes"
          className="episode-view-card"
          data-test="charView-episode"
        >
          {episodes.map((episode) => {
            return (
              <Link
                to={{
                  pathname: `/episodes/episode/${episode.id}`,
                  state: { data: episode },
                }}
                key={episode.id}
              >
                <Card.Grid style={gridStyle} key={episode.id}>
                  <h1>{episode.episode}</h1>
                  <h3>{episode.name}</h3>
                </Card.Grid>
              </Link>
            );
          })}
        </Card>
        <Card title="Recommended Episodes">
          {relevantRecommends.map((data) => {
            return (
              <Link
                to={{
                  pathname: `/episodes/episode/${data.id}`,
                  state: { data: data },
                }}
                key={data.id}
              >
                <Card.Grid style={gridStyle} key={data.id}>
                  <h1>{data.episode}</h1>
                  <h3>{data.name}</h3>
                </Card.Grid>
              </Link>
            );
          })}
        </Card>
      </div>
    );
  }
}
