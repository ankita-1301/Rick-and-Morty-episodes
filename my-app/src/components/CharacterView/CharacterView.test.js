import React from "react";
import Enzyme, { shallow } from "enzyme";
import CharacterView from "./CharacterView";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

//test setup for characterView component
const state = {
  state: {
    data: {
      id: 2,
      name: "Morty Smith",
      status: "Alive",
      species: "Human",
      type: "",
      gender: "Male",
      origin: {
        name: "Earth (C-137)",
        url: "https://rickandmortyapi.com/api/location/1",
      },
      location: {
        name: "Earth (Replacement Dimension)",
        url: "https://rickandmortyapi.com/api/location/20",
      },
      image: "https://rickandmortyapi.com/api/character/avatar/2.jpeg",
      episode: [
        "https://rickandmortyapi.com/api/episode/1",
        "https://rickandmortyapi.com/api/episode/2",
        "https://rickandmortyapi.com/api/episode/3",
      ],
      url: "https://rickandmortyapi.com/api/character/2",
      created: "2017-11-04T18:50:21.651Z",
    },
  },
};

const props = {};

test("CharacterView component snapshot", () => {
  const characterView = renderer.create(
    <CharacterView {...props} location={state} />
  );
  let charTree = characterView.toJSON();
  expect(charTree).toMatchSnapshot();
});

describe("Test character view component", () => {
  describe("Has rendered with mock props", () => {
    let component;
    beforeEach(() => {
      component = shallow(<CharacterView {...props} location={state} />);
    });

    it("Renders character view without error", () => {
      expect(component.length).toBe(1);
    });
  });
});
