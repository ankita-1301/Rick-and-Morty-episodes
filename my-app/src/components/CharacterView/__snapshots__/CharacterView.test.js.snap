// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters component snapshot 1`] = `
<div
  data-test="characterView-div"
>
  <h1>
    Morty Smith
  </h1>
  <span
    className="ant-avatar character-avatar ant-avatar-square ant-avatar-image ant-avatar-icon"
    style={
      Object {
        "fontSize": 100,
        "height": 200,
        "lineHeight": "200px",
        "width": 200,
      }
    }
  >
    <img
      onError={[Function]}
      src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
    />
  </span>
  <h5>
     Status : 
    Alive
  </h5>
  <h5>
     Species : 
    Human
  </h5>
  <h5>
     Gender : 
    Male
  </h5>
  <h5>
     Type : 
    
  </h5>
  <div
    className="ant-card episode-view-card ant-card-bordered"
    data-test="charView-episode"
  >
    <div
      className="ant-card-head"
      style={Object {}}
    >
      <div
        className="ant-card-head-wrapper"
      >
        <div
          className="ant-card-head-title"
        >
          Episodes
        </div>
      </div>
    </div>
    <div
      className="ant-card-body"
      style={Object {}}
    />
  </div>
  <div
    className="ant-card ant-card-bordered"
  >
    <div
      className="ant-card-head"
      style={Object {}}
    >
      <div
        className="ant-card-head-wrapper"
      >
        <div
          className="ant-card-head-title"
        >
          Recommended Episodes
        </div>
      </div>
    </div>
    <div
      className="ant-card-body"
      style={Object {}}
    />
  </div>
</div>
`;
